### PR TITLE
Column wise operator, filter and slice row

### PR DESCRIPTION
**Pull request by tiopramayudi**
**What this PR does / why we need it**:

In this PR, we add new capabilities in standard transformer:
* Conditional Update
* Column wise operator
* Filter row
* Slice row
**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

https://user-images.githubusercontent.com/2369255/161458901-ce538b34-7e28-49bb-b46a-bc82e064f5a4.mov

```release-note

```

**Checklist**

- [X] Added unit test, integration, and/or e2e tests
- [X] Tested locally
- [X] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

**Comment by pradithya** [comment]
Can you rebase?

**Comment by pradithya** [comment]
Can you check why the CI is failing? What's the video resolution? It seems very blurry, I am not sure if github perform compression but can you set the resolution to at least 720p?

File: api/pkg/transformer/pipeline/testdata/valid_table_transform_conditional_filtering.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          inputTable: driver_table
+          outputTable: transformed_driver_table
+          steps:
+            - updateColumns:
+              - column: customer_id
+                expression: customer_id
+              - column: driver_performa
+                conditions:
+                  - if: driver_table.Col("rating") * 2 <= 7
+                    expression: driver_table.Col("rating") * 1
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/testdata/valid_table_transform_conditional_filtering.yaml:26]: So this expression will only apply to rows that satisfy `driver_table.Col("rating") * 2 <= 7` condition, right? I now feel calling it `if` feel strange. Maybe `toRowsIf` sounds better?

Comment (pradithya) [reply api/pkg/transformer/pipeline/testdata/valid_table_transform_conditional_filtering.yaml:26]: or `rowSelector` ?

Comment (tiopramayudi) [reply api/pkg/transformer/pipeline/testdata/valid_table_transform_conditional_filtering.yaml:26]: `rowSelector` sounds better

File: api/pkg/transformer/pipeline/util.go [hunk 2]
```diff
@@ -16,6 +18,51 @@ func evalJSONPath(env *Environment, jsonPath string) (interface{}, error) {
...
 func evalExpression(env *Environment, expression string) (interface{}, error) {
+	val, err := getVal(env, expression)
+	if err != nil {
+		return nil, err
+	}
+	switch exprVal := val.(type) {
+	case *operation.OperationNode:
+		val, err = exprVal.Execute()
+	case operation.OperationNode:
+		val, err = exprVal.Execute()
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/util.go:29]: nit: Can it do it like this?

```go
case  *operation.OperationNode, operation.OperationNode:
    val, err = exprVal.Execute()
```

File: api/pkg/transformer/pipeline/util.go [hunk 3]
```diff
@@ -16,6 +18,51 @@ func evalJSONPath(env *Environment, jsonPath string) (interface{}, error) {
...
+	}
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/util.go:30]: What happen with the `default` case?

File: api/pkg/transformer/pipeline/util.go [hunk 4]
```diff
@@ -16,6 +18,51 @@ func evalJSONPath(env *Environment, jsonPath string) (interface{}, error) {
 }
 
 func evalExpression(env *Environment, expression string) (interface{}, error) {
+	val, err := getVal(env, expression)
+	if err != nil {
+		return nil, err
+	}
+	switch exprVal := val.(type) {
+	case *operation.OperationNode:
+		val, err = exprVal.Execute()
```
Comment (pradithya) [inline api/pkg/transformer/pipeline/util.go:27]: Is there any compilation penalty with this?

File: api/pkg/transformer/symbol/operator.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package symbol
+
+import (
+	"github.com/gojek/merlin/pkg/transformer/types/operation"
+)
+
+func (sr Registry) GreaterOp(left, right interface{}) interface{} {
```
Comment (pradithya) [inline api/pkg/transformer/symbol/operator.go:7]: Can you add doc for each method?

File: api/pkg/transformer/symbol/registry.go [hunk 6]
```diff
@@ -81,6 +82,18 @@ func (sr Registry) evalArg(arg interface{}) (interface{}, error) {
 		return cplJsonPath.LookupFromContainer(sr.jsonObjectContainer())
 	case *series.Series:
 		return val.GetRecords(), nil
+	case *operation.OperationNode:
```
Comment (pradithya) [inline api/pkg/transformer/symbol/registry.go:85]: nit: 
```go
case **operation.OperationNode, operation.OperationNode:
```

File: api/pkg/transformer/symbol/operator.go (hunk 5 above)
Comment (pradithya) [inline api/pkg/transformer/symbol/operator.go:7]: Do we have test for methods in operator.go?

File: api/pkg/transformer/types/operation/operation.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
+package operation
+
+import (
+	"fmt"
+
+	"github.com/gojek/merlin/pkg/transformer/types/series"
+)
+
```
Comment (pradithya) [inline api/pkg/transformer/types/operation/operation.go:8]: Can you add documentation too all exported symbols?

File: api/pkg/transformer/types/table/table_test.go [hunk 8]
```diff
@@ -771,3 +770,299 @@ func TestTable_NewRaw(t *testing.T) {
...
+			wantErr:    true,
+			errMessage: "failed slice col: A due to: slice index out of bounds",
+		},
+		{
+			name: "start < 0",
+			inputTable: New(
+				series.New([]int{1, 2, 3, 4, 5}, series.Int, "A"),
+				series.New([]string{"a", "b", "c", "d", "e"}, series.String, "B"),
+			),
+			startIdx:   -1,
```
Comment (pradithya) [inline api/pkg/transformer/types/table/table_test.go:838]: Can we allow negative index? It'll be useful for selecting series entry without having to know the series length. E.g.
```python
>>> a = [1,2,3,4,5,6,7,8,9,10]
# return last item
>>> a[-1]
10

# return 2nd last item
>>> a[-2]
9

# return last 2 items
>>> a[-2:]
[9, 10]
```

File: docs/user-guide/standard_transformer.md [hunk 9]
```diff
@@ -577,6 +577,85 @@ tableTransformation:
...
+        - column: "col2"
+          conditions:
+          - if: table1.Col('col1') * 2 > 10
+            expression: table1.Col('col1')
+          - default:
+              expression: -1
+```
+
+There are two ways to update columns:
+* Update column to the whole rows. Users need to only specify `column` and `expression`. Evaluation from that `expression` must contains of one value or array / series that has the same length with the rest of the column. Following the example:
```
Comment (pradithya) [inline docs/user-guide/standard_transformer.md:589]: Rephares

```text
Update all rows in the column. You need to specify `column`and `expression`. `column` determines which column to be updated and `expression` determines the value that will be used to update the column.
Values produced by the `expression` must be a scalar or a series that has the same length as the other columns. Following the example:
```

File: docs/user-guide/standard_transformer.md [hunk 10]
```diff
@@ -577,6 +577,85 @@ tableTransformation:
...
+  ```
+    - updateColumns:
+      - column: "customer_id"
+        expression: "cust_1" # the value is scalar and will be broadcasted to all the row
+      - column: "s2id"
+        expression: S2ID(table1.Col('lat'), table2.Col('lon'), 12) # the value is array or series that the length should be the same with the rest of the columns in a table 
+  ```
+* Update column to the subset of rows based on conditions. For this users can set multiple `if` with `expression` and also default value if none of conditions are match. For example users have following table
```
Comment (pradithya) [inline docs/user-guide/standard_transformer.md:597]: Rephrase

```text
Update subset of rows in the columns given some row selector condition. 
```

File: docs/user-guide/standard_transformer.md [hunk 11]
```diff
@@ -577,6 +577,85 @@ tableTransformation:
...
+tableTransformation:
+     inputTable: myTable
+     outputTable: myTransformedTable
+     steps:
+     - filterRow:
+        condition: myTable.Col('total_booking_1w') < 5
+```
+
+### Slice Row
+Slice row is an operation to slice a table based on start(lower bound) and end index(upper bound) that given by the user. The result will including start index but not end index. Below is the example of this operation
```
Comment (pradithya) [inline docs/user-guide/standard_transformer.md:650]: The result includes starting index but excluding end index.

File: ui/src/pages/version/components/forms/components/transformer/components/table_operations/UpdateColumnPanel.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+              description="Add column to updated or created"
+              onClick={() => onAddColumn()}
+            />
+          </EuiFlexItem>
+        </EuiFlexGroup>
+      </EuiDragDropContext>
+    </EuiPanel>
+  );
+
+  //   const tableColumns = [
```
Comment (pradithya) [inline ui/src/pages/version/components/forms/components/transformer/components/table_operations/UpdateColumnPanel.js:91]: Can we remove?

File: api/pkg/transformer/types/table/table_test.go (hunk 8 above)
Comment (tiopramayudi) [reply api/pkg/transformer/types/table/table_test.go:838]: can, I'll update this

File: api/pkg/transformer/pipeline/testdata/valid_table_transform_conditional_filtering.yaml (hunk 1 above)
Comment (karzuo) [reply api/pkg/transformer/pipeline/testdata/valid_table_transform_conditional_filtering.yaml:26]: sounds good. But `conditions` a bit not clear. Maybe call it `selectiveTransform` ?

File: api/pkg/transformer/pipeline/util.go (hunk 2 above)
Comment (tiopramayudi) [reply api/pkg/transformer/pipeline/util.go:29]: with the above `exprVal` type will be `interface` and can't call `Execute`

File: api/pkg/transformer/types/table/table_test.go (hunk 8 above)
Comment (tiopramayudi) [reply api/pkg/transformer/types/table/table_test.go:838]: # return 2nd last item
>>> a[-2]
what is the end index in this case ya?
>>> a[-2:]
for this case the end index is nil

File: api/pkg/transformer/pipeline/util.go (hunk 4 above)
Comment (tiopramayudi) [reply api/pkg/transformer/pipeline/util.go:27]: what do you mean by `compilation penalty`?

Comment (pradithya) [reply api/pkg/transformer/pipeline/util.go:27]: Nevermind, I thought `exprVal.Execute()` is calling expr `Execute`

File: ui/src/pages/version/components/forms/components/transformer/components/table_operations/SelectUpdateColumnStrategy.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    strategy !== "" ? option.value === strategy : option.value === ""
+  );
+
+  return (
+    <EuiForm>
+      <EuiFormRow
+        fullWidth
+        label={
+          <FormLabelWithToolTip
+            label="Choose Strategy *"
```
Comment (pradithya) [inline ui/src/pages/version/components/forms/components/transformer/components/table_operations/SelectUpdateColumnStrategy.js:37]: I think we can drop `Choose`.

File: api/pkg/transformer/types/table/table_test.go (hunk 8 above)
Comment (pradithya) [reply api/pkg/transformer/types/table/table_test.go]: I think you should mimic Python behavior

**Review by pradithya — APPROVED** [review]
LGTM!

**Pull request merged**